scripts: Add build-blocks-manifest command

mreishus opened the pull request:
## What?

Adds a new `wp-scripts build-blocks-manifest` command.

## Why?

To help plugin authors take advantage of the new function `wp_register_block_metadata_collection()` offered in https://core.trac.wordpress.org/changeset/59132 . It requires all block.json files to be compiled into one php manifest file, which this command does.

## How?

Looks for block.json files, keys them off of the parent directory they belong to, then passes to `json2php` (new dependency alert) to write to a php file.

## Testing Instructions

* **In the Gutenberg repo**: Run `npm run build:blocks-manifest` and inspect the `build/blocks-manifest.php` file.
* **As wp-scripts**: I'm not sure, this is my first time working on this, but I hope it will 'just work' due to the PR following the conventions.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

mreishus committed:
scripts: Add build-blocks-manifest command

mreishus commented:
Tag @felixarntz  ^ 
Is this what you had in mind? This is my first time working in wp-scripts.

felixarntz commented on packages/scripts/scripts/build-blocks-manifest.js:
```diff
@@ -0,0 +1,62 @@
+const { getArgFromCLI } = require( '../utils' );
+
+// Set default paths
+const defaultInputDir = 'build';
```
Should the default be `build` or `src`? I can see why you chose `build`, but a counter argument would be that this means you always have to run `npm run build` _before_ you run `npm run build:blocks-manifest`.

To my knowledge, the `block.json` files are copied as is from `src` to `build`, so in that case I think `src` is a better default. WDYT?

mreishus replied:
Hmm, I think I was influenced from my recent work in the WooCommerce plugin. That repo generates `block.json` files as a part of its main build process that are in directories named differently than their source values in `assets/`. Over there, the only way I could generate the right information was to pull from `build/` while also writing to `build/`. That may be atypical though.

mreishus replied:
I was about to move it to `src` here, but I found that in order to keep the `npm run build:blocks-manifest` command working in the root, I'd have to look through 4 src directories:
```
./packages/blocks/src/
./packages/widgets/src/
./packages/block-library/src/
./packages/edit-widgets/src/
```
We don't have a way to specify multiple input dirs. It may be the `build->build` after is a useful pattern for monorepos?

gziolo replied:
This one is interesting. When you use `wp-scripts build` then `block.json` files get copied from `src` to build and sometimes they even can get modified:

https://github.com/WordPress/gutenberg/blob/79940e0ca94cd814631e9a29f2d9178352246c70/packages/scripts/config/webpack.config.js#L314-L345

In effect, I think it's a valid approach to source the `block.json` from the `build` folder to put them in the `build` folder. In the long run, these probably should be an option on the `wp-scripts build --use-manifest`. However, we can get there in steps.

felixarntz replied:
That all makes sense, let's keep using `build` for now.

felixarntz reviewed (commented):
@mreishus Thanks for jumping on this to create the PR, it looks great to me so far. Kudos for already adding such extensive documentation! ❤️ 

I'm personally not the best candidate to review the JS command implementation, would be great to get reviews from people that were previously more involved in `@wordpress/scripts`. Maybe @gziolo you can help here?

gziolo commented on package.json:
```diff
@@ -272,6 +272,7 @@
 	"scripts": {
 		"build": "npm run build:packages && wp-scripts build",
 		"build:analyze-bundles": "npm run build -- --webpack-bundle-analyzer",
+		"build:blocks-manifest": "wp-scripts build-blocks-manifest",
```
This could go next in another PR to make it all simpler to land 😄

gziolo replied:
This could be removed for now as it isn’t connected anywhere.

mreishus replied:
Thank you, removed here: e24747ee85

sirreal reviewed (commented):
The code looks reasonable to me. It would be good to add a CHANGELOG entry.

A basic test would be nice, it could be a snapshot of the expected result. DEWP has some testing like this that could serve as inspiration:

https://github.com/WordPress/gutenberg/blob/trunk/packages/dependency-extraction-webpack-plugin/test/build.js

sirreal reviewed (approved):
I tested this out in Gutenberg and it seems to work well. I did list a couple nice-to-haves in [my previous review](https://github.com/WordPress/gutenberg/pull/65866#pullrequestreview-2357318949) but nothing blocking.

felixarntz reviewed (approved):
Giving this a 2nd approval, pending the minor feedback. Thanks @mreishus!

mreishus committed:
Remove package.json entry

mreishus committed:
add test

mreishus committed:
npm install in root to update lockfile

mreishus commented:
* added snapshot test 389b12858de3390d88fcad5f177ec0e590ffeca4 
* rebased and solved lockfile conflict (by dropping the npm install commit and making a new one)

mreishus committed:
Add changelog entry

jsnajdr commented on packages/scripts/scripts/build-blocks-manifest.js:
```diff
@@ -0,0 +1,62 @@
+const blockJsonFiles = glob( './**/block.json', {
+	cwd: path.resolve( process.cwd(), inputDir ),
+	absolute: true,
+} );
```
The script should first check if the `build` directory exists and abort with error if it doesn't.

It makes sense to run this script only if the blocks were already built and the `build` directory was created. If you forgot to run the build first, you should be told about it, instead of silently creating an empty manifest file.

mreishus replied:
:+1: Check https://github.com/WordPress/gutenberg/pull/65866/commits/e673f314c2f0cd3db49471c2892257889d4bff56

mreishus committed:
error on empty and missing dirs, update tests

felixarntz commented:
@jsnajdr Could you please take another look at this? Ideally, we could merge this before Thursday, as I am currently working on a dev note post for the change for WordPress 6.7, and it would make sense to have this available to reference for developers already in that post.

See https://make.wordpress.org/core/?p=115864&preview=1&_ppp=848634d99d for the public preview of that dev note. @gziolo @mreishus Could you please also review my draft and let me know your feedback?

gziolo commented:
> Ideally, we could merge this before Thursday, as I am currently working on a dev note post for the change for WordPress 6.7, and it would make sense to have this available to reference for developers already in that post.

Merging this PR is not enough to publish it to npm. The next package release will happen next week on Wednesday. Let me know if you would like to publish it separately on demand earlier.

> Could you please also review my draft and let me know your feedback?

I'm checking the draft next.

gziolo commented:
> See https://make.wordpress.org/core/?p=115864&preview=1&_ppp=848634d99d for the public preview of that dev note.

I'm not sure what is the best place for feedback. I can start here and move it elsewhere upon request:
- It needs to be clarified `wp_register_block_metadata_collection()` is an additional step on top of the existing registration process. In effect, I would include the snippet illustrating how the blocks get registered using:
    ```php
    wp_register_block_metadata_collection(
        WP_PLUGIN_DIR . '/my-block-library/build',
        WP_PLUGIN_DIR . '/my-block-library/build/blocks-manifest.php'
    );
    register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/pricing' );
    register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/services', array( 'render_callback' => 'my_block_library_render_callback' ) ); // It should still be possible to pass additional settings, so might be worth highlighting.
    register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/testimonial/block.json' ); // Does it work with full path, too?
   ```
- As noted in the code snippet, some notes about the path defined to the `block.json` file when using the collection would be helpful.
- Totally optional, illustrate that passing additional settings when registering the block is still compatible with the metadata collection.

mreishus commented:
```php
register_block_type( WP_PLUGIN_DIR . '/my-block-library/build/testimonial/block.json' ); // Does it work with full path, too?
```
Indeed, it works with full paths as well.

For the draft, it looks good! I agree with the "clarify it's an additional step" from @gziolo. For myself, I have no suggestions, other than optionally, you may want to also include https://make.wordpress.org/core/2022/10/07/improved-php-performance-for-core-blocks-registration/ when showing previous context.

Also, I'm sure you guys already know this, but just to be un-ambiguous, I don't have the ability to merge this PR. :)

gziolo commented:
Everything works as intended:

<img width="511" alt="Screenshot 2024-10-17 at 13 13 51" src="https://github.com/user-attachments/assets/6c0df930-f43e-4398-8597-6f742a52f476">

gziolo committed:
Merge branch 'trunk' into add/blocks-metadata

gziolo committed:
Update CHANGELOG.md

gziolo commented:
Once PR lands in `trunk`, I will cherry pick to `wp/latest` and trigger npm publishing.

jsnajdr reviewed (approved):
This looks good to me, thanks @mreishus for implementing the suggestion to abort right away when the `build` directory doesn't exist.

gziolo commented:
Publishing to npm with https://github.com/WordPress/gutenberg/actions/runs/11385894200.
